File: DataStructure/BST.cpp
#include<iostream>
using namespace std;

typedef struct node 
{
    int data;
    struct node *lchild;
    struct node *rchild; 
}NODE,*PNODE,**PPNODE;

void Insert(PPNODE Head,int no)
{
    PNODE newn = new NODE;
    newn->data = no;
    newn->lchild = NULL;
    newn->rchild = NULL;

    if(*Head == NULL)
    {
        *Head = newn;
    }
    else
    {
        PNODE temp = *Head;
        while(1)        // Unconditional loop       for(;;)
        {
            if(no == (temp->data))
            {
                delete newn;
                break;
            }
            else if(no > (temp->data))
            {
                if(temp->rchild == NULL)
                {
                    temp->rchild = newn;
                    break;
                }
                temp = temp->rchild;    
            }
            else if(no < (temp->data))
            {
                if(temp->lchild == NULL)
                {
                    temp->lchild = newn;
                    break;
                }
                temp = temp->lchild;
            }
        }
    }
}

bool Search(PNODE Head,int no)
{
    bool flag = false;

    if(Head == NULL)
    {
        return flag;
    }
    else
    {
        while(Head != NULL)
        {
            if(Head->data == no)
            {
                flag = true;
                break;
            }
            else if(no > Head->data)
            {
                Head = Head->rchild;
            }
            else if(no < Head->data)
            {
                Head = Head->lchild;
            }
        }
        return flag;
    }
}

int Count(PNODE Head)
{
    static int iCnt;
    if(Head != NULL)
    {
        iCnt++;
        Count(Head->lchild);
        Count(Head->rchild);
    }
    return iCnt;
}

int CountLeaf(PNODE Head)
{
    static int iCnt;
    if(Head != NULL)
    {
        if((Head->lchild == NULL)  && (Head->rchild == NULL))
        {
            iCnt++;
        }
        CountLeaf(Head->lchild);
        CountLeaf(Head->rchild);
    }
    return iCnt;
}

int CountParent(PNODE Head)
{
    static int iCnt;
    if(Head != NULL)
    {
        if((Head->lchild != NULL)  || (Head->rchild != NULL))
        {
            iCnt++;
        }
        CountParent(Head->lchild);
        CountParent(Head->rchild);
    }
    return iCnt;
}



int main()
{
    int no = 0,iret = 0;
    bool bret = false;
    PNODE first = NULL;
    
    Insert(&first,51);
    Insert(&first,21);
    Insert(&first,101);
    //Insert(&first,55);
    //Insert(&first,75);
    //Insert(&first,20);

    cout<<"Enter number to search\n";
    cin>>no;

    bret = Search(first,no);
    if(bret == true)
    {
        cout<<"Number is there\n";
    }
    else
    {
        cout<<"Number is not there\n";
    }

    iret = Count(first);
    cout<<"Elements inside tree : "<<iret<<"\n";

    iret = CountLeaf(first);
    cout<<"Number of leaf nodes are : "<<iret<<"\n";
    
    iret = CountParent(first);
    cout<<"Number of parent node are : "<<iret<<"\n";
    
    return 0;
}

File: DataStructure/BST.java
import java.util.*;

class Node
{
    public int data;
    public Node lchild;
    public Node rchild;
}
// 51   21  101  105  55  20 
class BinaryTree
{
    private Node Head;

    public BinaryTree()
    {
        Head = null;
    } 
    
    public void Insert(int no)
    {
        Node newn = new Node();
        newn.data = no;
        newn.lchild = null;
        newn.rchild = null;

        if(Head == null)
        {
            Head = newn;
        }
        else
        {
            Node temp = Head;
            while(true)     // unconditional loop
            {
                if(temp.data == no)
                {
                    temp = null;
                    break;
                }
                else if(no > temp.data)
                {
                    if(temp.rchild == null)
                    {
                        temp.rchild = newn;
                        break;
                    }
                    temp = temp.rchild;
                }
                else if(no < temp.data)
                {
                    if(temp.lchild == null)
                    {
                        temp.lchild = newn;
                        break;
                    }
                    temp = temp.lchild;
                }
            }
        }
    }

    public boolean Search(int no)
    {
        boolean flag = false;
        if(Head == null)
        {
            return flag;
        }
        else
        {
            while(Head != null)
            {
                if(Head.data == no)
                {
                    flag = true;
                    break;
                }
                else if(no > Head.data)
                {
                    Head = Head.rchild;
                }
                else if(no < Head.data)
                {
                    Head = Head.lchild;
                }
            }
        }
        return flag;
    }
}

class BST
{
    public static void main(String arr[])
    {
        boolean bret = false;
        int ino = 0,iret = 0;
        
        Scanner sobj = new Scanner(System.in);
        
        BinaryTree bobj = new BinaryTree();
        bobj.Insert(51);
        bobj.Insert(101);
        bobj.Insert(21);

        System.out.println("Enter number to search");
        ino = sobj.nextInt();
        bret = bobj.Search(ino);
        
        if(bret == true)
        {
            System.out.println("Number is there");
        }
        else
        {
            System.out.println("Number is not there");
        }
        iret = sobj.Count();
        System.out.println("Elements inside binary tree are "+iret);
    }
}